ql/src/java/org/apache/hadoop/hive/ql/optimizer/ConvertJoinMapJoin.java

```diff
@@ -657,22 +661,41 @@ private boolean convertJoinBucketMapJoin(JoinOperator joinOp, OptimizeTezProcCon
     boolean updatePartitionCols = false;
     List<Integer> positions = new ArrayList<>();
 
+    final CustomBucketFunction bucketFunction;
     if (listBucketCols.get(0).size() != bigTablePartitionCols.size()) {
       updatePartitionCols = true;
       // Prepare updated partition columns for small table(s).
       // Get the positions of bucketed columns
 
       int bigTableExprPos = 0;
       Map<String, ExprNodeDesc> colExprMap = bigTableRS.getColumnExprMap();
+      final boolean[] retainedColumns = new boolean[listBucketCols.get(0).size()];
       for (ExprNodeDesc bigTableExpr : bigTablePartitionCols) {
         // It is guaranteed there is only 1 list within listBucketCols.
-        for (String colName : listBucketCols.get(0)) {
+        for (int i = 0; i < listBucketCols.get(0).size(); i++) {
+          final String colName = listBucketCols.get(0).get(i);
           if (colExprMap.get(colName).isSame(bigTableExpr)) {
             positions.add(bigTableExprPos);
+            retainedColumns[i] = true;
           }
         }
         bigTableExprPos = bigTableExprPos + 1;
       }
+
+      Preconditions.checkState(opTraits.getCustomBucketFunctions().size() == 1);
+      if (opTraits.getCustomBucketFunctions().get(0) == null) {
+        bucketFunction = null;
```

Comment:
set default to null and avoid unnecessary conditional checks

Reply:
Done
https://github.com/apache/hive/pull/5409/commits/c6aa9d2d0a81ff2e8834e1350a18405e593fe212